src/python/zquantum/core/serialization.py

```diff
@@ -43,7 +59,45 @@ def iterencode(self, o: Any, _one_shot: bool = ...) -> Iterator[str]:
         return super().iterencode(preprocess(o))
 
 
+class OrquestraDecoder(json.JSONDecoder):
+    """Custom decoder for loading data dumped by ZapataEncoder."""
+
+    SCHEMA_MAP = {
+        "zapata-v1-value_estimate": ValueEstimate.from_dict,
+        "zapata-v1-optimization_result": lambda obj: optimization_result(**obj)
+    }
+
+    def __init__(self, *args, **kwargs):
+        json.JSONDecoder.__init__(
+            self, object_hook=self.object_hook, *args, **kwargs
+        )
+
+    def object_hook(self, obj):
+        # Parts of the below if-elif-else are sketchy, because for some objects there is
```

Comment:
I'd add a TODO here and appropriate story to the tracker.

Reply:
It would be the story about overhauling the serialization mechanism we were talking about earlier, right?

Reply:
Yeah, that's true 👍